src/app_client.rs

```diff
@@ -51,10 +56,55 @@ async fn process_block(
 				rpc_fetched.len()
 			);
 
-			let unfetched_count = positions.len() - ipfs_fetched.len() - rpc_fetched.len();
-			if unfetched_count > 0 {
-				return Err(anyhow!("Failed to fetch {} cells", unfetched_count));
+			let mut cells = vec![];
+			cells.extend(ipfs_fetched);
+			cells.extend(rpc_fetched);
+
+			if positions.len() > cells.len() {
+				return Err(anyhow!(
+					"Failed to fetch {} cells",
+					positions.len() - cells.len()
+				));
+			}
+
+			let count = verify_proof(
+				block.number,
+				block.dimensions.rows as u16,
+				block.dimensions.cols as u16,
+				cells.clone(),
+				block.commitment.clone(),
+			);
+			log::info!(
+				"Completed {} verification rounds for block {}",
+				count,
+				block.number
+			);
+
+			// TODO: If there are some invalid cells should we fail?
+
+			for cell in cells {
+				if let Err(error) = ipfs.insert(cell.clone().to_ipfs_block()) {
```

Comment:
Same as with sync, we should make sure that we're sending just the RPC retrieved cells, as not to spam the network with duplicates...

Reply:
Sure Ill change that, on both places.

Reply:
On three places actually.